tests.js

```diff
@@ -16,45 +16,167 @@
+    return Object.keys(this.tests).map((item) => (
+      [item, `${urlPrefix}/${item.replace(/\./g, '/')}`]
+    ));
+  }
+
+  listAllEndpoints(urlPrefix = '') {
+    return [
+      ...this.listMainEndpoints(urlPrefix),
+      ...this.listIndividual(urlPrefix)
+    ];
+  }
+
+  next(after) {
+    const afterURL = new URL(after);
+    if (!this.httpOnly && afterURL.protocol === 'http:') {
+      return `https://${this.host}${afterURL.pathname}`;
+    }
+    const endpoints = this.listMainEndpoints('/tests');
+    const index = endpoints.findIndex((item) => {
+      return item[1] === afterURL.pathname;
+    }) + 1;
+
+    if (index >= endpoints.length) {
+      return null;
+    }
+
+    const endpoint = endpoints[index][1];
+
+    if (this.endpoints[endpoint.replace('/tests', '')].httpsOnly) {
+      const newUrl = `https://${this.host}${endpoint}`;
+      if (this.httpOnly) {
+        // Skip this endpoint and go to the next
+        return this.next(newUrl);
       }
+      return newUrl;
+    }
+
+    return `http://${this.host}${endpoint}`;
+  }
+
+  getTests(endpoint) {
+    let idents = [];
+    if (endpoint in this.endpoints) {
+      idents = this.endpoints[endpoint].entries;
     } else {
-      begin = 0;
-      if (limit) {
-        end = limit;
-      }
+      idents = Object.keys(this.tests).filter(
+          (ident) => ident.startsWith(endpoint.substr(1).replace(/\//g, '.'))
+      );
+    }
+
+    const tests = {};
+
+    for (const ident of idents) {
+      tests[ident] = this.tests[ident];
     }
 
-    return this.items.slice(begin, end).map((item) => {
-      return `${item.protocol}://${this.host}${item.pathname}`;
-    });
+    return tests;
+  }
+
+  getScope(endpoint) {
+    const e = this.endpoints[endpoint];
+    return e ? e.scope : '';
   }
 
-  listIndividual() {
-    return Object.entries(this.individualItems)
-        .sort((a, b) => (a[0].localeCompare(b[0])));
+  generateTestPage(endpoint, testScope) {
+    const theseTests = this.getTests(endpoint);
+    const individual = !(endpoint in this.endpoints);
+
+    if (!testScope) {
+      testScope = individual ? '' : this.getScope(endpoint);
+    }
+
+    const lines = [
+      '<!DOCTYPE html>',
+      '<html>',
+      '<head>',
+      '<meta charset="utf-8">',
+      '<script src="/resources/json3.min.js"></script>',
+      '<script src="/resources/harness.js"></script>',
+      '<script src="/resources/core.js"></script>',
+      '<style>',
+      '@media (prefers-color-scheme: dark) {',
+      'body {',
+      'background-color: #111;',
+      'color: white;',
+      '}',
+      '}',
+      '</style>',
+      '</head>',
+      '<body>',
+      '<p id="status">Running tests...</p>',
+      '<script>'
+    ];
+
+    for (const [ident, test] of Object.entries(theseTests)) {
+      for (const scope of test.scope) {
+        if (!testScope || scope == testScope) {
+          lines.push(`bcd.addTest("${ident}", ${JSON.stringify(test.tests)}, "${scope}");`);
```

Comment:
This is a place where I expect resolving https://github.com/foolip/mdn-bcd-collector/issues/376 will change things a lot. Rather than feeding input to the client-side test generator here, we should just output stuff like:
```js
bcd.addTest('api.Document', function() {
  return 'Document' in window;
});
```

But not in this PR :)